java/vector/src/main/java/org/apache/arrow/vector/stream/MessageSerializer.java

```diff
@@ -98,26 +100,41 @@ public static Schema deserializeSchema(ReadChannel in) throws IOException {
   }
 
   /**
-   * Serializes an ArrowRecordBatch.
+   * Serializes an ArrowRecordBatch. Returns the offset and length of the written batch.
    */
-  public static long serialize(WriteChannel out, ArrowRecordBatch batch)
+  public static ArrowBlock serialize(WriteChannel out, ArrowRecordBatch batch)
       throws IOException {
     long start = out.getCurrentPosition();
     int bodyLength = batch.computeBodyLength();
-
     ByteBuffer metadata = WriteChannel.serialize(batch);
+
+    int messageLength = 4 + metadata.remaining() + bodyLength;
     ByteBuffer serializedHeader =
-        serializeHeader(MessageHeader.RecordBatch, bodyLength + metadata.remaining() + 4);
+        serializeHeader(MessageHeader.RecordBatch, messageLength);
```

Comment:
The `header` union could be arbitrarily large since it grows with the number of fields in the record batch or schema